From bb39005bf9ca6c25421edaa97454bbb076f898ff Mon Sep 17 00:00:00 2001
From: Thao Luong <thao.luong.uw@rvc.renesas.com>
Date: Wed, 27 May 2015 11:46:45 +0700
Subject: [PATCH 25/25] omxvideodec: Add hacks to choose default mode for 
 omxh264dec

omxh264dec support 3 modes to send data to downstream:
 - copy mode: send data to downstream using memcpy
 - no-copy mode: send data to downstream using userptr
 - dmabuf mode: send data to downstream using dmabuf method
Normally, default mode is dmabuf mode. This patch will support
choosing default mode through setting in gstomx.conf

To choose copy mode as default mode, add "use-copy-mode-as-defalut" into
hacks in gstomx.conf
Ex:
[omxh264dec]
...
hacks=no-disable-outport;default-pix-aspect-ratio;use-copy-mode-as-default

To choose no-copy mode as default mode, add "use-userptr-mode-as-default"
into hacks in gstomx.conf
Ex:
[omxh264dec]
...
hacks=no-disable-outport;default-pix-aspect-ratio;use-userptr-mode-as-default

If add both "use-copy-mode-as-default" and use-userptr-mode-as-default"
It will be invalid value, the normal mode (dmabuf) will be choosen

refs #69206
Signed-off-by: Thao Luong <thao.luong.uw@rvc.renesas.com>
---
 omx/gstomx.c         |  4 ++++
 omx/gstomx.h         | 10 ++++++++++
 omx/gstomxvideodec.c | 19 +++++++++++++++++++
 3 files changed, 33 insertions(+)

diff --git a/omx/gstomx.c b/omx/gstomx.c
index 4d75fa2..e1ca49c 100644
--- a/omx/gstomx.c
+++ b/omx/gstomx.c
@@ -2470,6 +2470,10 @@ gst_omx_parse_hacks (gchar ** hacks)
       hacks_flags |= GST_OMX_HACK_NOT_HANDLE_CODEC_DATA;
     else if (g_str_equal (*hacks, "default-pix-aspect-ratio"))
       hacks_flags |= GST_OMX_HACK_DEFAULT_PIXEL_ASPECT_RATIO;
+    else if (g_str_equal (*hacks, "use-copy-mode-as-default"))
+      hacks_flags |= GST_OMX_HACK_USE_COPY_MODE_AS_DEFAULT;
+    else if (g_str_equal (*hacks, "use-userptr-mode-as-default"))
+      hacks_flags |= GST_OMX_HACK_USE_USERPTR_MODE_AS_DEFAULT;
     else
       GST_WARNING ("Unknown hack: %s", *hacks);
     hacks++;
diff --git a/omx/gstomx.h b/omx/gstomx.h
index 636a800..dc3a888 100644
--- a/omx/gstomx.h
+++ b/omx/gstomx.h
@@ -129,6 +129,16 @@ G_BEGIN_DECLS
  */
 #define GST_OMX_HACK_DEFAULT_PIXEL_ASPECT_RATIO                       G_GUINT64_CONSTANT (0x0000000000000400)
 
+/* omxh264dec support 3 modes (copy, no-copy(userptr), dmabuf) to choose.
+ * Default is dmabuf mode. This hack will choose copy mode is default mode
+ */
+#define GST_OMX_HACK_USE_COPY_MODE_AS_DEFAULT                       G_GUINT64_CONSTANT (0x0000000000000800)
+
+/* omxh264dec support 3 modes (copy, no-copy(userptr), dmabuf) to choose.
+ * Default is dmabuf mode. This hack will choose no-copy(userptr) mode is default mode
+ */
+#define GST_OMX_HACK_USE_USERPTR_MODE_AS_DEFAULT                       G_GUINT64_CONSTANT (0x0000000000001000)
+
 typedef struct _GstOMXCore GstOMXCore;
 typedef struct _GstOMXPort GstOMXPort;
 typedef enum _GstOMXPortDirection GstOMXPortDirection;
diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index 242eb09..b831fb7 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -202,6 +202,7 @@ gst_omx_video_dec_init (GstOMXVideoDec * self)
 
   g_mutex_init (&self->drain_lock);
   g_cond_init (&self->drain_cond);
+
   self->no_copy = FALSE;
 #ifdef HAVE_MMNGRBUF
   self->use_dmabuf = TRUE;
@@ -310,6 +311,24 @@ gst_omx_video_dec_open (GstVideoDecoder * decoder)
 
   GST_DEBUG_OBJECT (self, "Opened EGL renderer");
 #endif
+  /* Use hacks to choose default mode, normally default mode is dmabuf */
+  if (!(klass->cdata.hacks & GST_OMX_HACK_USE_COPY_MODE_AS_DEFAULT && 
+        klass->cdata.hacks & GST_OMX_HACK_USE_USERPTR_MODE_AS_DEFAULT))
+  {
+    if (klass->cdata.hacks & GST_OMX_HACK_USE_COPY_MODE_AS_DEFAULT)
+    {
+      GST_DEBUG_OBJECT (self, "Use copy mode as default");
+      self->no_copy = FALSE;
+      self->use_dmabuf = FALSE;
+    }
+    if (klass->cdata.hacks & GST_OMX_HACK_USE_USERPTR_MODE_AS_DEFAULT)
+    {
+      GST_DEBUG_OBJECT (self, "Use no-copy (userptr) mode as default");
+      self->no_copy = TRUE;
+      self->use_dmabuf = FALSE;
+    }
+  } else 
+    GST_DEBUG_OBJECT (self, "Use dmabuf mode as default");
 
   return TRUE;
 }
-- 
2.4.0

